shell/components/mic-input.js

```diff
@@ -0,0 +1,141 @@
+/*
+@license
+Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+*/
+
+import Xen from './xen/xen.js';
+import Icons from './icons.css.js';
+
+const log = Xen.logFactory('MicInput', 'blue');
+
+class MicInput extends Xen.Base {
+  _getInitialState() {
+    return {
+    };
+  }
+  _didMount() {
+    if (this._supportsSpeech()) {
+      this._initSpeech();
+      document.addEventListener('keydown', this._onKeydown.bind(this));
+      document.addEventListener('keyup', this._onKeyup.bind(this));
+    }
+  }
+  _update(props, state) {
+    if (state.active && !state.recognizing) {
+      this._start();
+    }
+    if (!state.active && state.recognizing) {
+      this._stop();
+    }
+  }
+  _render(props, state) {
+    return state;
+  }
+  _supportsSpeech() {
+    return ('webkitSpeechRecognition' in window);
+  }
+  _initSpeech(state) {
+    const recognition = new window.webkitSpeechRecognition();
+    recognition.continuous = true;
+    recognition.interimResults = true;
+    recognition.onstart = this._onRecognitionStart.bind(this);
+    recognition.onerror = this._onRecognitionError.bind(this);
+    recognition.onend = this._onRecognitionEnd.bind(this);
+    recognition.onresult = this._onRecognitionResult.bind(this);
+    this._setState({recognition});
+  }
+  start() {
+    const {recognition, recognizing, finalTranscript} = this._state;
+    if (recognition) {
+      this._start();
+    } // else complain
+  }
+  _start() {
+    const {recognition, recognizing, duration, finalTranscript} = this._state;
+    if (recognizing) {
+      recognition.stop();
+    } else {
+      recognition.start();
+      this._setState({
+        finalTranscript: '',
+        ignoreOnEnd: false
+      });
+    }
+  }
+  _stop() {
+    const {recognition, recognizing, duration, finalTranscript} = this._state;
+    if (recognizing) {
+      recognition.stop();
+    }
+  }
+  _onStartClick() {
+    this.start();
+  }
+  _onRecognitionStart() {
+    this._setState({recognizing: true});
+    this._fire('start');
+  }
+  _onRecognitionError() {
+    if (event.error == 'no-speech') {
+      // TODO: something?
+    }
+    if (event.error == 'audio-capture') {
+      // TODO: something?
+    }
+    if (event.error == 'not-allowed') {
+      // TODO: something?
+    }
+    this._setState({ignoreOnEnd: true});
+  }
+  _onRecognitionResult() {
+    let {ignoreOnEnd, finalTranscript} = this._state;
+    let interimTranscript = '';
+    for (let i = event.resultIndex; i < event.results.length; ++i) {
```

Comment:
This loop is interesting, mostly in the initialization state. Is it that speech events continue to get concatenated to the results array, but the system lets us know what we've seen already via `resultIndex`?

Reply:
Mostly this was cribbed from example code, I didn't study it thoroughly. But yes, it looks like each event provides addenda to the `results[]` and `resultIndex` says where the new bits start.
Text in `interimTranscript` can change by the next event (as the engine has more time to cogitate on it [maybe?], or more input comes in). What is provided from these events is fungible.